File: python/agentpin/crypto.py
"""ECDSA P-256 cryptographic operations for AgentPin."""

import base64
import hashlib

from cryptography.hazmat.primitives import hashes, serialization
from cryptography.hazmat.primitives.asymmetric import ec
from cryptography.hazmat.primitives.asymmetric.ec import (
    ECDSA,
    SECP256R1,
    EllipticCurvePrivateKey,
    EllipticCurvePublicKey,
)


def generate_key_pair() -> tuple:
    """Generate a new ECDSA P-256 keypair.

    Returns:
        Tuple of (private_key_pem: str, public_key_pem: str)
    """
    private_key = ec.generate_private_key(SECP256R1())
    private_key_pem = private_key.private_bytes(
        encoding=serialization.Encoding.PEM,
        format=serialization.PrivateFormat.PKCS8,
        encryption_algorithm=serialization.NoEncryption(),
    ).decode("utf-8")

    public_key_pem = private_key.public_key().public_bytes(
        encoding=serialization.Encoding.PEM,
        format=serialization.PublicFormat.SubjectPublicKeyInfo,
    ).decode("utf-8")

    return private_key_pem, public_key_pem


def sign_data(private_key_pem: str, data: bytes) -> str:
    """Sign data with a PEM-encoded private key. Returns base64-encoded DER signature."""
    private_key = serialization.load_pem_private_key(private_key_pem.encode("utf-8"), password=None)
    assert isinstance(private_key, EllipticCurvePrivateKey)
    signature = private_key.sign(data, ECDSA(hashes.SHA256()))
    return base64.b64encode(signature).decode("utf-8")


def verify_signature(public_key_pem: str, data: bytes, signature_b64: str) -> bool:
    """Verify a base64-encoded DER signature against data using a PEM public key."""
    try:
        public_key = serialization.load_pem_public_key(public_key_pem.encode("utf-8"))
        assert isinstance(public_key, EllipticCurvePublicKey)
        sig_bytes = base64.b64decode(signature_b64)
        public_key.verify(sig_bytes, data, ECDSA(hashes.SHA256()))
        return True
    except Exception:
        return False


def generate_key_id(public_key_pem: str) -> str:
    """Generate a key ID: SHA-256 of DER-encoded SPKI, hex-encoded."""
    public_key = serialization.load_pem_public_key(public_key_pem.encode("utf-8"))
    der_bytes = public_key.public_bytes(
        encoding=serialization.Encoding.DER,
        format=serialization.PublicFormat.SubjectPublicKeyInfo,
    )
    return sha256_hex(der_bytes)


def sha256_hash(data: bytes) -> bytes:
    """SHA-256 hash of arbitrary data."""
    return hashlib.sha256(data).digest()


def sha256_hex(data: bytes) -> str:
    """SHA-256 hash, hex-encoded."""
    return hashlib.sha256(data).hexdigest()

File: javascript/src/crypto.js
/**
 * ECDSA P-256 cryptographic operations for AgentPin.
 * Uses Node.js built-in crypto module — zero external dependencies.
 */

import { createSign, createVerify, generateKeyPairSync, createHash, createPublicKey } from 'crypto';

/**
 * Generate a new ECDSA P-256 keypair.
 * @returns {{ privateKeyPem: string, publicKeyPem: string }}
 */
export function generateKeyPair() {
    const { privateKey, publicKey } = generateKeyPairSync('ec', {
        namedCurve: 'prime256v1',
        publicKeyEncoding: { type: 'spki', format: 'pem' },
        privateKeyEncoding: { type: 'pkcs8', format: 'pem' },
    });
    return { privateKeyPem: privateKey, publicKeyPem: publicKey };
}

/**
 * Sign data with a PEM-encoded private key. Returns base64-encoded DER signature.
 * @param {string} privateKeyPem
 * @param {Buffer|Uint8Array} data
 * @returns {string} Base64-encoded DER signature
 */
export function signData(privateKeyPem, data) {
    const sign = createSign('SHA256');
    sign.update(data);
    const signature = sign.sign(privateKeyPem);
    return signature.toString('base64');
}

/**
 * Verify a base64-encoded DER signature against data using a PEM public key.
 * @param {string} publicKeyPem
 * @param {Buffer|Uint8Array} data
 * @param {string} signatureB64 - Base64-encoded DER signature
 * @returns {boolean}
 */
export function verifySignature(publicKeyPem, data, signatureB64) {
    try {
        const verify = createVerify('SHA256');
        verify.update(data);
        const signature = Buffer.from(signatureB64, 'base64');
        return verify.verify(publicKeyPem, signature);
    } catch {
        return false;
    }
}

/**
 * Generate a key ID from a public key PEM: SHA-256 of DER-encoded SPKI, hex-encoded.
 * @param {string} publicKeyPem
 * @returns {string} Hex-encoded SHA-256 hash
 */
export function generateKeyId(publicKeyPem) {
    const keyObject = createPublicKey(publicKeyPem);
    const der = keyObject.export({ type: 'spki', format: 'der' });
    return sha256Hex(der);
}

/**
 * SHA-256 hash of arbitrary data.
 * @param {Buffer|Uint8Array|string} data
 * @returns {Buffer}
 */
export function sha256Hash(data) {
    return createHash('sha256').update(data).digest();
}

/**
 * SHA-256 hash, hex-encoded.
 * @param {Buffer|Uint8Array|string} data
 * @returns {string}
 */
export function sha256Hex(data) {
    return createHash('sha256').update(data).digest('hex');
}
